Bluejay/Bluejay/ConnectionObserver.swift

```diff
@@ -30,13 +30,18 @@ public protocol ConnectionObserver: class {
     
     /// Called whenever a peripheral is disconnected.
     func disconnected(from peripheral: Peripheral, with error: Error?)
+    
+    /// Called before Blujay tries to reconnect from an accidental disconnection from a peirpheral
+    func shouldReconnect(to: Peripheral) -> Bool
 }
 
 /// Slightly less gross way to make the ConnectionObserver protocol's functions optional.
 extension ConnectionObserver {
     public func bluetoothAvailable(_ available: Bool) {}
     public func connected(to peripheral: Peripheral) {}
     public func disconnected(from peripheral: Peripheral, with error: Error?) {}
+    public func shouldReconnect(to: Peripheral) -> Bool { return true }
```

Comment:
Hello @JeremyChiang, I'm sorry but the work isn't done yet. I thought it was more simple, but there are a lot of  corner cases that still must be handled. I still didn't test it yet and probably there are still some "ideas" that I left mistakenly in the code. I was really busy at work and since the last commit I didn't had time to work again on this issue.
But I think it could be a great time to clarify what's on my mind.
The first startegy I tried to apply was on connection observers, unfortunately this could never work, primary because if there is more than one connection observer active as listeners is impossible to decide who must decide if it should reconnect or not. What you are still seeing in the code is a misprint about this old approach that I must remember to remove.
The second strategy I applied is the one with the `reconnectionHandler`, this one is set on Bluejay instance itself. I'm pretty confident that this should work, but I have some concern about cancellation: I'm not totally sure that if I cancel a connection `CBCentralManager` will notify `didDisconnectPeripheral` or `didFailToConnect` by returning an `Error` and making the user with the `reconnectionHandler ` apply its logic as such an event occur. As I said I still need to test it.
About the `BluejayConfiguration` I didn't know about the #165 , totally correct about what you say at point 3, but point 2 I think is just a matter of taste, I applied this patter since also in Apple frameworks is appearing more often, for instance in `UIGraphicsImageRenderer`, you can pass a format that specifies some options and Apple provides some defaults values as well, but no big deal about.
I hope to start testing it next week.